File: backend/fief/schemas/role.py
from typing import List, Optional

from pydantic import UUID4

from fief.schemas.generics import BaseModel, CreatedUpdatedAt, UUIDSchema
from fief.schemas.permission import PermissionEmbedded


class RoleCreate(BaseModel):
    name: str
    granted_by_default: bool
    permissions: List[UUID4]


class RoleUpdate(BaseModel):
    name: Optional[str]
    granted_by_default: Optional[bool]
    permissions: Optional[List[UUID4]]


class BaseRole(UUIDSchema, CreatedUpdatedAt):
    name: str
    granted_by_default: bool


class Role(BaseRole):
    permissions: List[PermissionEmbedded]


class RoleEmbedded(BaseRole):
    pass

File: admin-frontend/src/schemas/role.ts
import { CreatedUpdatedAt, UUIDSchema } from './generics';
import { PermissionEmbedded } from './permission';

export interface RoleCreate {
  name: string;
  granted_by_default: boolean;
  permissions: string[];
}

export interface RoleUpdate {
  name?: string
  granted_by_default?: boolean;
  permissions?: string[];
}

interface BaseRole extends UUIDSchema, CreatedUpdatedAt {
  name: string;
  granted_by_default: boolean;
}

export interface Role extends BaseRole {
  permissions: PermissionEmbedded[];
}

export interface RoleEmbedded extends BaseRole { }
